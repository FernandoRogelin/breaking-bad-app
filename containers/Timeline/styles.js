import styled from 'styled-components';

import ButtonComponent from '#/components/Button';

export const Section = styled.section`
  display: flex;
  align-items: center;
  flex-direction: column;
`;

export const Title = styled.h1`
  margin: 0;
`;

export const Itens = styled.div`
  display: flex;
  margin: 0 15px;
  padding: 6px 15px;
  justify-content: space-between;

  &:not(:last-child) {
    border-bottom: solid 1px #ddd;
  }

  @media only screen and (max-width: 440px) {
    flex-direction: column;
  }
`;

export const Name = styled.p`
  margin: 0;
  color: #8c9aa3;
  font-size: 20px;
`;

export const Button = styled(ButtonComponent)`
  padding: 2px 6px;
  font-size: 12px;

  &:not(:last-child) {
    margin-right: 10px;
  }
`;

export const Actions = styled.div``;

export const Top = styled.div`
  display: flex;
  padding: 8px 114px 8px 30px;
  justify-content: space-between;
  border-bottom: 1px solid #919191;
`;

export const Header = styled.header`
  width: 60%;
  display: flex;
  justify-content: flex-end;
`;

export const Text = styled.p`
  margin: 0;

  @media only screen and (max-width: 440px) {
    &:last-child {
      display: none;
    }
  }
`;
