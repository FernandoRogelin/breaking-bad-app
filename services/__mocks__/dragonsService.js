export const getDragons = jest.fn().mockResolvedValue({
  status: 200,
  data: [
    {
      id: '54',
      createdAt: '2019-04-13T18:42:21.806Z',
      name: 'Ysera',
      type: 'Verde',
      histories: [],
    },
    {
      id: '56',
      createdAt: '2019-04-14T07:31:43.898Z',
      name: 'Rosamond',
      type: 'azure',
      histories: [],
    },
    {
      id: '57',
      createdAt: '2019-04-13T23:44:08.488Z',
      name: 'Leopoldo',
      type: 'white',
      histories: [],
    },
    {
      id: '61',
      createdAt: '2019-04-14T09:40:39.780Z',
      name: 'Sadie',
      type: 'orange',
      histories: [],
    },
    {
      id: '62',
      createdAt: '19/05/2018',
      name: 'Dinquedo',
      type: 'Amarelo',
      histories: [],
    },
  ],
});

export const deleteDragon = jest.fn().mockResolvedValue({
  status: 200,
});

export const getDetailsDragon = jest.fn().mockResolvedValue({
  status: 200,
  data: {
    id: '62',
    createdAt: '19/05/2018',
    name: 'Dinquedo',
    type: 'Amarelo',
    histories: [],
  },
});
