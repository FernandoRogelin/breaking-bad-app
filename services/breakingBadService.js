import axios from 'axios';

export const urlBase = 'https://www.breakingbadapi.com/api/';

export const getCharacters = () => axios.get(`${urlBase}characters`);

export const getCharacterDetails = id =>
  axios.get(`${urlBase}characters/${id}`);

export const getEpisodesList = () => axios.get(`${urlBase}episodes`);

export const getEpisodeDetails = id => axios.get(`${urlBase}episodes/${id}`);

export const getQuotes = () => axios.get(`${urlBase}quotes`);
