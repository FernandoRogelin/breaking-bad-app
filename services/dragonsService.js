import axios from 'axios';

export const urlBase =
  'http://5c4b2a47aa8ee500142b4887.mockapi.io/api/v1/dragon';

export const getDragons = () => axios.get(urlBase);

export const getDetailsDragon = id => axios.get(`${urlBase}/${id}`);

export const createDragon = ({ name, type, createdAt }) =>
  axios.post(urlBase, {
    name,
    type,
    createdAt,
  });

export const deleteDragon = id => axios.delete(`${urlBase}/${id}`);

export const editDragon = (id, { name, type, createdAt }) =>
  axios.put(`${urlBase}/${id}`, {
    name,
    type,
    createdAt,
  });
