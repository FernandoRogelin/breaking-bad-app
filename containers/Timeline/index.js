import React, { useState, useEffect } from 'react';

import { Link } from '#/server/routes';

import { getDragons, deleteDragon } from '#/services/dragonsService';
import Content from '#/components/Content';

import {
  Section,
  Title,
  Itens,
  Name,
  Button,
  Actions,
  Header,
  Text,
  Top,
} from './styles';

export default function Timeline() {
  const [dragons, setDragons] = useState();
  const [called, setCalled] = useState(false);

  async function handleDragons() {
    try {
      const { data } = await getDragons();
      setDragons(data);
      setCalled(true);
    } catch (e) {
      console.log(e);
    }
  }

  async function handleDelete(id) {
    try {
      await deleteDragon(id);
      location.reload();
    } catch (e) {
      console.log(e);
    }
  }

  useEffect(() => {
    if (!called) {
      handleDragons();
    }
  }, [called]);

  return (
    <Section>
      <Title>Lista de dragões</Title>
      <Header>
        <Link route="create">
          <Button>Adicionar</Button>
        </Link>
      </Header>
      {dragons && (
        <Content>
          <Top>
            <Text>Nome</Text>
            <Text>Ações</Text>
          </Top>
          {dragons
            .sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0))
            .map(({ name, id }) => (
              <Itens key={id}>
                <Name>{name}</Name>
                <Actions>
                  <Link route="edit" params={{ id }}>
                    <Button>Editar</Button>
                  </Link>
                  <Button onClick={() => handleDelete(id)}>Remover</Button>
                  <Link route="details" params={{ id }}>
                    <Button>Detalhes</Button>
                  </Link>
                </Actions>
              </Itens>
            ))}
        </Content>
      )}
    </Section>
  );
}
