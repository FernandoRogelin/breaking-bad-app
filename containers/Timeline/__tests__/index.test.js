import React from 'react';
import { render, wait, fireEvent } from 'react-testing-library';

import { getDragons, deleteDragon } from '#/services/dragonsService';
import Timeline from '..';

jest.mock('#/services/dragonsService');

describe('Timeline container', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('Timeline snapshot', () => {
    const { container } = render(<Timeline />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('Inicial flow', async () => {
    const { getByTestId } = render(<Timeline />);

    await wait(() => expect(getDragons).toHaveBeenCalledTimes(1));

    fireEvent.click(getByTestId('delete-button'));

    await wait(() => expect(deleteDragon).toHaveBeenCalledTimes(1));
  });
});
