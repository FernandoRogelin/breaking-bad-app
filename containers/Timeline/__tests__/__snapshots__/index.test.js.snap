// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline container Timeline snapshot 1`] = `
<section
  class="styles__Section-sc-150p1d9-0 iUVPOe"
>
  <h1
    class="styles__Title-sc-150p1d9-1 IdGEn"
  >
    Lista de drag√µes
  </h1>
  <header
    class="styles__Header-sc-150p1d9-7 gITzGW"
  >
    <button
      class="Button-n3cyj4-0 styles__Button-sc-150p1d9-4 jBEnXM"
      data-testid="add-button"
    >
      Adicionar
    </button>
  </header>
</section>
`;
